File: AlgoritmosOrden.java
// Ordene a mesma lista utilizando Shell Sort, Merge Sort, Selection Sort, Quick Sort, Bucket Sort e Radix Sort. 
//Registre os tempos de execução e número de comparações realizadas.

import java.util.*;

public class AlgoritmosOrden {

    public static void main(String[] args) {
        int[] sizes = {1000, 5000, 10000}; // Tamanhos diferentes de listas
        for (int size : sizes) {
            int[] arr = generateRandomArray(size);

            System.out.println("Tamanho da Lista: " + size);

            // Shell Sort
            int[] shellArr = arr.clone();
            long startTime = System.nanoTime();
            int shellComparisons = shellSort(shellArr);
            long endTime = System.nanoTime();
            long shellTime = endTime - startTime;
            System.out.printf("Shell Sort -> Tempo: %d ns, Comparações: %d\n", shellTime, shellComparisons);

            // Merge Sort
            int[] mergeArr = arr.clone();
            int[] mergeComparisons = {0};
            startTime = System.nanoTime();
            mergeSort(mergeArr, 0, size - 1, mergeComparisons);
            endTime = System.nanoTime();
            long mergeTime = endTime - startTime;
            System.out.printf("Merge Sort -> Tempo: %d ns, Comparações: %d\n", mergeTime, mergeComparisons[0]);

            // Selection Sort
            int[] selectionArr = arr.clone();
            startTime = System.nanoTime();
            int selectionComparisons = selectionSort(selectionArr);
            endTime = System.nanoTime();
            long selectionTime = endTime - startTime;
            System.out.printf("Selection Sort -> Tempo: %d ns, Comparações: %d\n", selectionTime, selectionComparisons);

            // Quick Sort
            int[] quickArr = arr.clone();
            int[] quickComparisons = {0};
            startTime = System.nanoTime();
            quickSort(quickArr, 0, size - 1, quickComparisons);
            endTime = System.nanoTime();
            long quickTime = endTime - startTime;
            System.out.printf("Quick Sort -> Tempo: %d ns, Comparações: %d\n", quickTime, quickComparisons[0]);

            // Bucket Sort
            int[] bucketArr = arr.clone();
            startTime = System.nanoTime();
            int bucketComparisons = bucketSort(bucketArr, size);
            endTime = System.nanoTime();
            long bucketTime = endTime - startTime;
            System.out.printf("Bucket Sort -> Tempo: %d ns, Comparações: %d\n", bucketTime, bucketComparisons);

            // Radix Sort
            int[] radixArr = arr.clone();
            startTime = System.nanoTime();
            int radixComparisons = radixSort(radixArr, size);
            endTime = System.nanoTime();
            long radixTime = endTime - startTime;
            System.out.printf("Radix Sort -> Tempo: %d ns, Comparações: %d\n", radixTime, radixComparisons);

            System.out.println();
        }
    }

    public static int[] generateRandomArray(int size) {
        Random rand = new Random();
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = rand.nextInt(size);
        }
        return arr;
    }

    public static int shellSort(int[] arr) {
        int n = arr.length;
        int comparisons = 0;

        for (int gap = n / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
                int temp = arr[i];
                int j;
                for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                    arr[j] = arr[j - gap];
                    comparisons++;
                }
                arr[j] = temp;
                comparisons++;
            }
        }
        return comparisons;
    }

    public static void mergeSort(int[] arr, int left, int right, int[] comparisons) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(arr, left, mid, comparisons);
            mergeSort(arr, mid + 1, right, comparisons);
            merge(arr, left, mid, right, comparisons);
        }
    }

    public static void merge(int[] arr, int left, int mid, int right, int[] comparisons) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] leftArray = new int[n1];
        int[] rightArray = new int[n2];

        for (int i = 0; i < n1; ++i) leftArray[i] = arr[left + i];
        for (int j = 0; j < n2; ++j) rightArray[j] = arr[mid + 1 + j];

        int i = 0, j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            comparisons[0]++;
            if (leftArray[i] <= rightArray[j]) {
                arr[k] = leftArray[i];
                i++;
            } else {
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            arr[k] = leftArray[i];
            i++;
            k++;
            comparisons[0]++;
        }

        while (j < n2) {
            arr[k] = rightArray[j];
            j++;
            k++;
            comparisons[0]++;
        }
    }

    public static int selectionSort(int[] arr) {
        int n = arr.length;
        int comparisons = 0;

        for (int i = 0; i < n - 1; i++) {
            int min_idx = i;
            for (int j = i + 1; j < n; j++) {
                comparisons++;
                if (arr[j] < arr[min_idx]) {
                    min_idx = j;
                }
            }
            int temp = arr[min_idx];
            arr[min_idx] = arr[i];
            arr[i] = temp;
            comparisons++;
        }
        return comparisons;
    }

    public static void quickSort(int[] arr, int low, int high, int[] comparisons) {
        if (low < high) {
            int pi = partition(arr, low, high, comparisons);
            quickSort(arr, low, pi - 1, comparisons);
            quickSort(arr, pi + 1, high, comparisons);
        }
    }

    public static int partition(int[] arr, int low, int high, int[] comparisons) {
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            comparisons[0]++;
            if (arr[j] < pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        comparisons[0]++;
        return i + 1;
    }

    public static int bucketSort(int[] arr, int n) {
        if (n <= 0) return 0;

        List<Integer>[] buckets = new List[10];
        for (int i = 0; i < buckets.length; i++) {
            buckets[i] = new ArrayList<>();
        }

        for (int num : arr) {
            int bucketIndex = num / (n / 10);
            buckets[bucketIndex].add(num);
        }

        int comparisons = 0;
        int index = 0;
        for (List<Integer> bucket : buckets) {
            Collections.sort(bucket);
            for (int num : bucket) {
                arr[index++] = num;
                comparisons++;
            }
        }
        return comparisons;
    }

    public static int radixSort(int[] arr, int n) {
        int m = getMax(arr, n);
        int comparisons = 0;

        for (int exp = 1; m / exp > 0; exp *= 10) {
            comparisons += countSort(arr, n, exp);
        }
        return comparisons;
    }

    public static int getMax(int[] arr, int n) {
        int max = arr[0];
        for (int i = 1; i < n; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }
        return max;
    }

    public static int countSort(int[] arr, int n, int exp) {
        int[] output = new int[n];
        int[] count = new int[10];
        int comparisons = 0;

        for (int i = 0; i < n; i++) {
            count[(arr[i] / exp) % 10]++;
        }

        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            output[count[(arr[i] / exp) % 10] - 1] = arr[i];
            count[(arr[i] / exp) % 10]--;
            comparisons++;
        }

        for (int i = 0; i < n; i++) {
            arr[i] = output[i];
        }
        return comparisons;
    }
}

File: AlgoritmosOrden.py
#Ordene a mesma lista utilizando Shell Sort, Merge Sort, Selection Sort, Quick Sort, Bucket Sort e Radix Sort. 
#Registre os tempos de execução e número de comparações realizadas.

import time
import random
import numpy as np

class SortingAlgorithms:
    def __init__(self):
        self.comparisons = 0

    def shell_sort(self, arr):
        self.comparisons = 0
        n = len(arr)
        gap = n // 2
        while gap > 0:
            for i in range(gap, n):
                temp = arr[i]
                j = i
                while j >= gap and arr[j - gap] > temp:
                    self.comparisons += 1
                    arr[j] = arr[j - gap]
                    j -= gap
                arr[j] = temp
                self.comparisons += 1
            gap //= 2
        return arr

    def merge_sort(self, arr):
        if len(arr) > 1:
            mid = len(arr) // 2
            L = arr[:mid]
            R = arr[mid:]
            self.merge_sort(L)
            self.merge_sort(R)
            i = j = k = 0
            while i < len(L) and j < len(R):
                self.comparisons += 1
                if L[i] < R[j]:
                    arr[k] = L[i]
                    i += 1
                else:
                    arr[k] = R[j]
                    j += 1
                k += 1
            while i < len(L):
                arr[k] = L[i]
                i += 1
                k += 1
            while j < len(R):
                arr[k] = R[j]
                j += 1
                k += 1
        return arr

    def selection_sort(self, arr):
        self.comparisons = 0
        n = len(arr)
        for i in range(n):
            min_idx = i
            for j in range(i+1, n):
                self.comparisons += 1
                if arr[j] < arr[min_idx]:
                    min_idx = j
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
        return arr

    def quick_sort(self, arr, low, high):
        if low < high:
            pi = self.partition(arr, low, high)
            self.quick_sort(arr, low, pi - 1)
            self.quick_sort(arr, pi + 1, high)
        return arr

    def partition(self, arr, low, high):
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            self.comparisons += 1
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1

    def bucket_sort(self, arr):
        self.comparisons = 0
        max_value = max(arr)
        n = len(arr)
        num_buckets = int(np.ceil(np.sqrt(n)))  # Ajuste para calcular o número de baldes

        buckets = [[] for _ in range(num_buckets)]

        for num in arr:
            index = min(num // (max_value // num_buckets + 1), num_buckets - 1)  # Garantir que o índice está no intervalo
            buckets[index].append(num)

        sorted_arr = []
        for bucket in buckets:
            sorted_arr.extend(sorted(bucket))
            self.comparisons += len(bucket)
        return sorted_arr

    def radix_sort(self, arr):
        self.comparisons = 0
        max_val = max(arr)
        exp = 1
        while max_val // exp > 0:
            self.count_sort(arr, exp)
            exp *= 10
        return arr

    def count_sort(self, arr, exp):
        n = len(arr)
        output = [0] * n
        count = [0] * 10
        for i in range(n):
            index = arr[i] // exp
            count[index % 10] += 1
        for i in range(1, 10):
            count[i] += count[i - 1]
        i = n - 1
        while i >= 0:
            index = arr[i] // exp
            output[count[index % 10] - 1] = arr[i]
            count[index % 10] -= 1
            i -= 1
        for i in range(n):
            arr[i] = output[i]
        self.comparisons += n

def measure_time_and_comparisons(sort_function, arr):
    start_time = time.time()
    sort_alg = SortingAlgorithms()
    sort_alg.comparisons = 0
    sorted_arr = sort_function(sort_alg, arr.copy())
    return time.time() - start_time, sort_alg.comparisons

sizes = [100, 1000, 10000]
results = {"Algorithm": ["Shell Sort", "Merge Sort", "Selection Sort", "Quick Sort", "Bucket Sort", "Radix Sort"]}

for size in sizes:
    results[f"{size} Size (Time)"] = []
    results[f"{size} Size (Comparisons)"] = []

sorting_algorithms = [
    lambda sort_alg, arr: sort_alg.shell_sort(arr),
    lambda sort_alg, arr: sort_alg.merge_sort(arr),
    lambda sort_alg, arr: sort_alg.selection_sort(arr),
    lambda sort_alg, arr: sort_alg.quick_sort(arr, 0, len(arr) - 1),
    lambda sort_alg, arr: sort_alg.bucket_sort(arr),
    lambda sort_alg, arr: sort_alg.radix_sort(arr),
]

for size in sizes:
    arr = np.random.randint(0, size * 10, size).tolist()
    for sort_function in sorting_algorithms:
        time_taken, comparisons = measure_time_and_comparisons(sort_function, arr)
        results[f"{size} Size (Time)"].append(time_taken)
        results[f"{size} Size (Comparisons)"].append(comparisons)

print(f"{'Algorithm':<20} {'100 Size (Time)':<20} {'100 Size (Comparisons)':<25} {'1000 Size (Time)':<20} {'1000 Size (Comparisons)':<25} {'10000 Size (Time)':<20} {'10000 Size (Comparisons)':<25}")
for i in range(len(results["Algorithm"])):
    print(f"{results['Algorithm'][i]:<20} {results['100 Size (Time)'][i]:<20.10f} {results['100 Size (Comparisons)'][i]:<25} {results['1000 Size (Time)'][i]:<20.10f} {results['1000 Size (Comparisons)'][i]:<25} {results['10000 Size (Time)'][i]:<20.10f} {results['10000 Size (Comparisons)'][i]:<25}")
